coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangeSplitPlanner.scala

```diff
@@ -0,0 +1,118 @@
+  *                                    is not "now" because of the delay in population of downsampled data
+  *                                    via spark job. If job is run every 6 hours,
+  *                                    `(System.currentTimeMillis - 12.hours.toMillis)`
+  *                                    may a function that could be passed. 12 hours to account for failures/reruns etc.
+  * @param splitThresholdMs the function that will provide the threshold in milliseconds. If the query time ranges is
+  *                         beyond this threshold, then the query will be split into smaller time range queries.
+  * @param splitSizeMs  If the query is to be broken then it will be of this size
+  * @param stitchDispatcher function to get the dispatcher for the stitch exec plan node
+  */
+class LongTimeRangeSplitPlanner(rawClusterPlanner: QueryPlanner,
+                                downsampleClusterPlanner: QueryPlanner,
+                                earliestRawTimestampFn: => Long,
+                                latestDownsampleTimestampFn: => Long,
+                                splitThresholdMs: => Long,
+                                splitSizeMs: => Long,
+                                stitchDispatcher: => PlanDispatcher) extends QueryPlanner with StrictLogging {
+
+  //Most of the logic is derived from LongTimeRangePlanner
+  def materialize(logicalPlan: LogicalPlan, qContext: QueryContext): ExecPlan = {
+    logicalPlan match {
+      case p: PeriodicSeriesPlan =>
+        val earliestRawTime = earliestRawTimestampFn
+        val offsetMillis = LogicalPlanUtils.getOffsetMillis(logicalPlan)
+        val lookbackMs = LogicalPlanUtils.getLookBackMillis(logicalPlan)
+        val startWithOffsetMs = p.startMs - offsetMillis
+        val endWithOffsetMs = p.endMs - offsetMillis
+        if (!logicalPlan.isRoutable) //can be answered wholly by raw cluster
+          splitPlan(p, qContext, p1 => rawClusterPlanner.materialize(p1, qContext))
+        else if (endWithOffsetMs < earliestRawTime) // can be answered wholly by downsample cluster
+          splitPlan(p, qContext, p1 => downsampleClusterPlanner.materialize(p1, qContext))
+        else if (startWithOffsetMs - lookbackMs >= earliestRawTime) //can be answered wholly by raw cluster
+          splitPlan(p, qContext, p1 => rawClusterPlanner.materialize(p1, qContext))
+        else if (endWithOffsetMs - lookbackMs < earliestRawTime) { //can be answered wholly by downsample cluster
+          val lastDownsampleSampleTime = latestDownsampleTimestampFn
+          val downsampleLp = if (endWithOffsetMs < lastDownsampleSampleTime) {
+            logicalPlan
+          } else {
+            copyWithUpdatedTimeRange(logicalPlan,
+              TimeRange(p.startMs, latestDownsampleTimestampFn + offsetMillis))
+          }
+          splitPlan(downsampleLp, qContext, p1 =>  downsampleClusterPlanner.materialize(p1, qContext))
+        } else {
```

Comment:
All of this copied code from LongTimeRangePlanner would go away if the split is done in SingleClusterPlanner.

Reply:
@vishramachandran moved logic inside SingleClusterPlanner